src/caffe/layers/cudnn_conv_layer.cpp

```diff
@@ -167,19 +233,228 @@ void CuDNNConvolutionLayer<Dtype>::Reshape(
 }
 
 template <typename Dtype>
+void CuDNNConvolutionLayer<Dtype>::GetConvAlgo(
+    const vector<Blob<Dtype>*>& bottom,
+    const vector<Blob<Dtype>*>& top,
+    const size_t workspace_bytes) {
+
+  for (int i = 0; i < bottom.size(); i++) {
+    // Get forward and backward algorithms
+    CUDNN_CHECK(cudnnGetConvolutionForwardAlgorithm(Caffe::cudnn_handle(),
+        bottom_descs_[i], filter_desc_, conv_descs_[i], top_descs_[i],
+        CUDNN_CONVOLUTION_FWD_SPECIFY_WORKSPACE_LIMIT,
+        workspace_bytes, &fwd_algo_[i]));
+    // Get backward algorithm for filter
+    CUDNN_CHECK(cudnnGetConvolutionBackwardFilterAlgorithm(
+        Caffe::cudnn_handle(),
+        bottom_descs_[i], top_descs_[i], conv_descs_[i], filter_desc_,
+        CUDNN_CONVOLUTION_BWD_FILTER_SPECIFY_WORKSPACE_LIMIT,
+        workspace_bytes, &bwd_filter_algo_[i]));
+    // Get backward algorithm for data
+    CUDNN_CHECK(cudnnGetConvolutionBackwardDataAlgorithm(
+        Caffe::cudnn_handle(),
+        filter_desc_, top_descs_[i], conv_descs_[i], bottom_descs_[i],
+        CUDNN_CONVOLUTION_BWD_DATA_SPECIFY_WORKSPACE_LIMIT,
+        workspace_bytes, &bwd_data_algo_[i]));
+  }
+}
+
+template <typename Dtype>
+void CuDNNConvolutionLayer<Dtype>::FindExConvAlgo(
+    const vector<Blob<Dtype>*>& bottom,
+    const vector<Blob<Dtype>*>& top,
+    const size_t workspace_bytes) {
+
+  // Number of algorithms we want to consider
+  // Since we only consider one algorithm (the fastest), set this to 1
+  const int kRequestAlgoCount = 1;
+  int fwd_algo_count;
+  int filter_algo_count;
+  int data_algo_count;
+
+  cudnnConvolutionFwdAlgoPerf_t       fwd_results[kRequestAlgoCount];
+  cudnnConvolutionBwdFilterAlgoPerf_t bwd_filter_results[kRequestAlgoCount];
+  cudnnConvolutionBwdDataAlgoPerf_t   bwd_data_results[kRequestAlgoCount];
+
+  // Allocate temporary buffer for weights used for backward filter FindEx
+  void *tmp_weights;
+  const int tmp_weights_size = sizeof(Dtype) * weight_offset_;
+  GPUMemoryManager::allocate(&tmp_weights, tmp_weights_size);
+
+  // workspace_bytes is the amount of available memory before allocating
+  // tmp_weights. So, size of tmp_weights should be subtracted from
+  // workspace_bytes to represent the correct amount of available memory.
+  if (!workspace.try_reserve(workspace_bytes - tmp_weights_size)) {
+    workspace.reserve(workspace_bytes - tmp_weights_size);
+  }
+
+  for (int i = 0; i < bottom.size(); i++) {
+    // Find forward algorithm
+    CUDNN_CHECK(cudnnFindConvolutionForwardAlgorithmEx(
```

Comment:
It should return identical results for every `i`, shouldn't it? May be better to run it one time only? ANd then populate `fwd_algo` array? Would be quite a big saving...